# 2014222 - Semester 2 CA-01 -April 2024¶

# Part 2

# Step 1: Import Python Libraries

In [34]:
# Utilise Python programming language in order to comply with the requisites of the assessment and perform adequate Machine
# Learning algorithms to discover and deliver insights.

# Import the necessary libraries (Pandas) in order to perform data cleansing. These are the libraries that are 
# conventionally used as a common practice in order to perform mathematical and statistical operations during a data analysis
# process

import pandas as pd

# Import the Plotly library in order to perform data visualisation procedures

import plotly.express as px
import plotly.graph_objects as go

# I am using this line of code to see all columns in a wide DataFrame

pd.set_option('display.max_columns', None)

# Using train_test_split and StandardScaler together in a machine learning pipeline
# ensures that the model is trained on a well-prepared dataset,
# leading to better performance and more reliable predictions.

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Using Sequential to create a model and Dense for adding layers is a powerful
# and straightforward way to build neural networks in TensorFlow.
# This approach is widely used for a plethora of deep learning tasks where
# complex architectures are not required, providing a great balance between simplicity and functionality.

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Step 2: Reading Dataset

In [2]:
# Read the CSV file by applying the pd.read_csv() function.

nba_data = pd.read_csv('NBA Dataset CSV C1.csv')

Reference to the dataset:

https://stats.nba.com/stats/leagueLeaders?LeagueID=00&PerMode=Totals&Scope=S&Season=2012-13&SeasonType=Regular%20Season&StatCategory=PTS

# Analyzing the Data

In [3]:
# Print the NBA table dataframe

nba_data

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0     2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1     2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2     2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3     2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4     2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   
...       ...               ...        ...   ...              ...         ...   
7288  2021-22          Playoffs    1629006   206      Josh Okogie  1610612750   
7289  2021-22          Playoffs    1630556   206  Kessler Edwards  1610612751   
7290  2021-22          Playoffs    1630201   206    Malachi Flynn  1610612761   
7291  2021-22          Playoffs     202693   206  Markieff Morris  1610612748   
7292  2021-22          Playoffs     200794   206     Paul Millsap  1610612755   

     TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0     OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1     LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2     MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3     HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4     NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   
...   ...  ..   ...  ...   ...     ...   ...   ...      ...  ...  ...     ...   
7288  MIN   1     2    0     0   0.000     0     0    0.000    0    0   0.000   
7289  BKN   2     7    0     0   0.000     0     0    0.000    0    0   0.000   
7290  TOR   6    36    0     7   0.000     0     3    0.000    0    0   0.000   
7291  MIA   2     3    0     1   0.000     0     0    0.000    0    0   0.000   
7292  PHI   1     6    0     0   0.000     0     0    0.000    0    0   0.000   

      OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  
0       46   594  640  374  116  105  280  143  2280  2462     1.34     0.41  
1       66   367  433  469  106   25  287  173  2133  1921     1.63     0.37  
2       97   513  610  551  129   67  226  110  2036  2446     2.44     0.57  
3       62   317  379  455  142   38  295  178  2023  1872     1.54     0.48  
4      134   326  460  171   52   32  175  205  1920  1553     0.98     0.30  
...    ...   ...  ...  ...  ...  ...  ...  ...   ...   ...      ...      ...  
7288     0     0    0    0    0    0    0    0     0     0     0.00     0.00  
7289     0     0    0    1    1    0    1    3     0     1     1.00     1.00  
7290     2     1    3    3    1    0    1    6     0    -1     3.00     1.00  
7291     0     1    1    0    0    0    1    2     0    -1     0.00     0.00  
7292     0     1    1    1    0    0    0    1     0     2     0.00     0.00  

[7293 rows x 30 columns]

In [4]:
# Print the dimensions of the NBA DataFrame.

nba_data.shape

(7293, 30)

As we can see that the NBA dataframe contains 7293 rows × 30 columns

In [5]:
# Print the first 5 rows of the NBA dataframe by applying the .head() method,
# This method is will display the top 5 observations of the dataset

nba_data.head()

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0  2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1  2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2  2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3  2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4  2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   

  TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0  OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1  LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2  MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3  HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4  NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   

   OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  
0    46   594  640  374  116  105  280  143  2280  2462     1.34     0.41  
1    66   367  433  469  106   25  287  173  2133  1921     1.63     0.37  
2    97   513  610  551  129   67  226  110  2036  2446     2.44     0.57  
3    62   317  379  455  142   38  295  178  2023  1872     1.54     0.48  
4   134   326  460  171   52   32  175  205  1920  1553     0.98     0.30

In [6]:
# Print the last 5 rows of the NBA dataframe by applying the .tail() method,
# Will display the last 5 observations of the dataset

nba_data.tail()

Year Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID TEAM  \
7288  2021-22    Playoffs    1629006   206      Josh Okogie  1610612750  MIN   
7289  2021-22    Playoffs    1630556   206  Kessler Edwards  1610612751  BKN   
7290  2021-22    Playoffs    1630201   206    Malachi Flynn  1610612761  TOR   
7291  2021-22    Playoffs     202693   206  Markieff Morris  1610612748  MIA   
7292  2021-22    Playoffs     200794   206     Paul Millsap  1610612755  PHI   

      GP  MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  \
7288   1    2    0    0     0.0     0     0      0.0    0    0     0.0     0   
7289   2    7    0    0     0.0     0     0      0.0    0    0     0.0     0   
7290   6   36    0    7     0.0     0     3      0.0    0    0     0.0     2   
7291   2    3    0    1     0.0     0     0      0.0    0    0     0.0     0   
7292   1    6    0    0     0.0     0     0      0.0    0    0     0.0     0   

      DREB  REB  AST  STL  BLK  TOV  PF  PTS  EFF  AST_TOV  STL_TOV  
7288     0    0    0    0    0    0   0    0    0      0.0      0.0  
7289     0    0    1    1    0    1   3    0    1      1.0      1.0  
7290     1    3    3    1    0    1   6    0   -1      3.0      1.0  
7291     1    1    0    0    0    1   2    0   -1      0.0      0.0  
7292     1    1    1    0    0    0   1    0    2      0.0      0.0

In [7]:
# Getting information about the NBA dataframe by applying the .info() method,
# Will display number of records in each column, data having null or not null, Data type,
# memory usage of the dataset
# helps to understand the data type and information about data

nba_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7293 entries, 0 to 7292
Data columns (total 30 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         7293 non-null   object 
 1   Season_type  7293 non-null   object 
 2   PLAYER_ID    7293 non-null   int64  
 3   RANK         7293 non-null   int64  
 4   PLAYER       7293 non-null   object 
 5   TEAM_ID      7293 non-null   int64  
 6   TEAM         7293 non-null   object 
 7   GP           7293 non-null   int64  
 8   MIN          7293 non-null   int64  
 9   FGM          7293 non-null   int64  
 10  FGA          7293 non-null   int64  
 11  FG_PCT       7293 non-null   float64
 12  FG3M         7293 non-null   int64  
 13  FG3A         7293 non-null   int64  
 14  FG3_PCT      7293 non-null   float64
 15  FTM          7293 non-null   int64  
 16  FTA          7293 non-null   int64  
 17  FT_PCT       7293 non-null   float64
 18  OREB         7293 non-null   int64  
 19  DREB  

Seems to be no missing values.

This information can be helpful for diagnosing issues with data processing and understanding the structure of the NBA dataset.


# Check for Duplication

In [8]:
# Getting several unique values in each columnan the data description the NBA dataframe
# by applying the .nunique() method,
# Will display continuous and categorical columns in the data.
# Duplicated data can be handled or removed based on further analysis
# helps to understand the data type and information about data

nba_data.nunique()

Year             10
Season_type       2
PLAYER_ID      1386
RANK            537
PLAYER         1382
TEAM_ID          30
TEAM             31
GP               83
MIN            2331
FGM             670
FGA            1210
FG_PCT          483
FG3M            261
FG3A            566
FG3_PCT         379
FTM             428
FTA             517
FT_PCT          539
OREB            285
DREB            590
REB             748
AST             544
STL             165
BLK             175
TOV             289
PF              276
PTS            1449
EFF            1567
AST_TOV         445
STL_TOV         226
dtype: int64

In [9]:
# Identifying null values in the NBA dataframe by applying the .isnull() method,

nba_data.isnull()

Year  Season_type  PLAYER_ID   RANK  PLAYER  TEAM_ID   TEAM     GP  \
0     False        False      False  False   False    False  False  False   
1     False        False      False  False   False    False  False  False   
2     False        False      False  False   False    False  False  False   
3     False        False      False  False   False    False  False  False   
4     False        False      False  False   False    False  False  False   
...     ...          ...        ...    ...     ...      ...    ...    ...   
7288  False        False      False  False   False    False  False  False   
7289  False        False      False  False   False    False  False  False   
7290  False        False      False  False   False    False  False  False   
7291  False        False      False  False   False    False  False  False   
7292  False        False      False  False   False    False  False  False   

        MIN    FGM    FGA  FG_PCT   FG3M   FG3A  FG3_PCT    FTM    FTA  \
0     False  False  False   False  False  False    False  False  False   
1     False  False  False   False  False  False    False  False  False   
2     False  False  False   False  False  False    False  False  False   
3     False  False  False   False  False  False    False  False  False   
4     False  False  False   False  False  False    False  False  False   
...     ...    ...    ...     ...    ...    ...      ...    ...    ...   
7288  False  False  False   False  False  False    False  False  False   
7289  False  False  False   False  False  False    False  False  False   
7290  False  False  False   False  False  False    False  False  False   
7291  False  False  False   False  False  False    False  False  False   
7292  False  False  False   False  False  False    False  False  False   

      FT_PCT   OREB   DREB    REB    AST    STL    BLK    TOV     PF    PTS  \
0      False  False  False  False  False  False  False  False  False  False   
1      False  False  False  False  False  False  False  False  False  False   
2      False  False  False  False  False  False  False  False  False  False   
3      False  False  False  False  False  False  False  False  False  False   
4      False  False  False  False  False  False  False  False  False  False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
7288   False  False  False  False  False  False  False  False  False  False   
7289   False  False  False  False  False  False  False  False  False  False   
7290   False  False  False  False  False  False  False  False  False  False   
7291   False  False  False  False  False  False  False  False  False  False   
7292   False  False  False  False  False  False  False  False  False  False   

        EFF  AST_TOV  STL_TOV  
0     False    False    False  
1     False    False    False  
2     False    False    False  
3     False    False    False  
4     False    False    False  
...     ...      ...      ...  
7288  False    False    False  
7289  False    False    False  
7290  False    False    False  
7291  False    False    False  
7292  False    False    False  

[7293 rows x 30 columns]

As we can see the NBA dataframe null values are false, meaning there is not missing values

In [10]:
# Identifying missing values in the NBA dataframe by applying the .isna().sum() methods,
# I am using this to get the number of missing records in each column

nba_data.isnull().sum()

Year           0
Season_type    0
PLAYER_ID      0
RANK           0
PLAYER         0
TEAM_ID        0
TEAM           0
GP             0
MIN            0
FGM            0
FGA            0
FG_PCT         0
FG3M           0
FG3A           0
FG3_PCT        0
FTM            0
FTA            0
FT_PCT         0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
PF             0
PTS            0
EFF            0
AST_TOV        0
STL_TOV        0
dtype: int64

As we can see the NBA dataframe does not contain any duplicates.

In [11]:
# Identifying the percentage of missing values in the NBA dataframe by applying the .isna().sum() methods,
# I am using this to get the number of missing records in each column

(nba_data.isnull().sum()/(len(nba_data)))* 100

Year           0.0
Season_type    0.0
PLAYER_ID      0.0
RANK           0.0
PLAYER         0.0
TEAM_ID        0.0
TEAM           0.0
GP             0.0
MIN            0.0
FGM            0.0
FGA            0.0
FG_PCT         0.0
FG3M           0.0
FG3A           0.0
FG3_PCT        0.0
FTM            0.0
FTA            0.0
FT_PCT         0.0
OREB           0.0
DREB           0.0
REB            0.0
AST            0.0
STL            0.0
BLK            0.0
TOV            0.0
PF             0.0
PTS            0.0
EFF            0.0
AST_TOV        0.0
STL_TOV        0.0
dtype: float64

As we can see the percentage of missing values for the columns is 0%

In [12]:
# Checking if there any missing or null values in the NBA dataframe by applying the .isna() method,

# and the .sum() method to return the sum of the duplicate values in the whole dataframe.

# The duplicated() method would return a series of 'True' or 'False', based on whether a row contains
# a duplicate value or not. Whereas, by adding the .sum() method, which
# is conventionally used to obtain the sum of a series of values,
# it is possible to obtain the whole sum of all the duplicated
# values contained in the dataframe

nba_data.isna().sum()

Year           0
Season_type    0
PLAYER_ID      0
RANK           0
PLAYER         0
TEAM_ID        0
TEAM           0
GP             0
MIN            0
FGM            0
FGA            0
FG_PCT         0
FG3M           0
FG3A           0
FG3_PCT        0
FTM            0
FTA            0
FT_PCT         0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
PF             0
PTS            0
EFF            0
AST_TOV        0
STL_TOV        0
dtype: int64

# Step 3: Data Reduction

In [13]:
# Print 10 random records of the NBA table dataframe, by applying the .sample() method
# I want to know which columns or variables can be dropped if they are not useful.

nba_data.sample(10)

Year       Season_type  PLAYER_ID  RANK             PLAYER  \
2356  2015-16  Regular%20Season     203112   295         Quincy Acy   
1199  2013-14          Playoffs     101162    41      Marcin Gortat   
5947  2020-21  Regular%20Season    1628964   255           Mo Bamba   
3059  2016-17  Regular%20Season     201155   306     Rodney Stuckey   
794   2013-14  Regular%20Season       2210   118  Richard Jefferson   
49    2012-13  Regular%20Season     201609    50       Goran Dragic   
6732  2021-22  Regular%20Season     203085   262      Austin Rivers   
6447  2020-21          Playoffs    1629730   214      Rayjon Tucker   
7020  2021-22  Regular%20Season    1628238   547         Paris Bass   
6821  2021-22  Regular%20Season    1629649   350   Ignas Brazdeikis   

         TEAM_ID TEAM  GP   MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  \
2356  1610612758  SAC  59   876  119  214   0.556    19    49    0.388   50   
1199  1610612764  WAS  11   382   58  118   0.492     0     0    0.000   27   
5947  1610612753  ORL  46   725  142  301   0.472    38   118    0.322   45   
3059  1610612754  IND  39   696   91  244   0.373    19    60    0.317   80   
794   1610612762  UTA  82  2213  294  654   0.450   123   301    0.409  120   
49    1610612756  PHX  77  2581  401  906   0.443    88   276    0.319  244   
6732  1610612743  DEN  67  1480  146  350   0.417    79   231    0.342   32   
6447  1610612755  PHI   1     2    0    0   0.000     0     0    0.000    1   
7020  1610612756  PHX   2     7    2    6   0.333     0     2    0.000    2   
6821  1610612753  ORL  42   536   81  188   0.431    27    87    0.310   21   

      FTA  FT_PCT  OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  \
2356   68   0.735    65   123  188   27   29   24   27  103   307   435   
1199   41   0.659    36    73  109   16    5   14   20   22   143   193   
5947   66   0.682    76   189  265   35   14   58   36   87   367   523   
3059  107   0.748     6    78   84   84   16    0   53   44   281   232   
794   162   0.741    18   201  219  130   56   14   94  168   831   754   
49    326   0.748    60   178  238  569  124   26  212  214  1134  1292   
6732   44   0.727    19    95  114   90   51    6   54  121   403   394   
6447    2   0.500     0     0    0    0    0    0    0    1     1     0   
7020    2   1.000     2     2    4    0    1    0    2    2     6     5   
6821   32   0.656    23    49   72   37   10    3   17   48   210   197   

      AST_TOV  STL_TOV  
2356     1.00     1.07  
1199     0.80     0.25  
5947     0.97     0.39  
3059     1.58     0.30  
794      1.38     0.60  
49       2.68     0.58  
6732     1.67     0.94  
6447     0.00     0.00  
7020     0.00     0.50  
6821     2.18     0.59

In [14]:
# Converts the column called 'Year' from object to integer by applying the 'astype()' method
# and pass 'int' to represent integers.

nba_data['Year'].str[:4].astype(int)

0       2012
1       2012
2       2012
3       2012
4       2012
        ... 
7288    2021
7289    2021
7290    2021
7291    2021
7292    2021
Name: Year, Length: 7293, dtype: int64

The dataype inside the column called 'Year'  now represents **integers.**

In [15]:
# Now that we converted year to represent integers
# I can create a new column called 'season_start_year'

nba_data['season_start_year'] = nba_data['Year'].str[:4].astype(int)

In [16]:
# Checking if the column season_start_year was successfully created
# by applying the .head() method,

nba_data.head()

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0  2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1  2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2  2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3  2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4  2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   

  TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0  OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1  LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2  MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3  HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4  NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   

   OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  \
0    46   594  640  374  116  105  280  143  2280  2462     1.34     0.41   
1    66   367  433  469  106   25  287  173  2133  1921     1.63     0.37   
2    97   513  610  551  129   67  226  110  2036  2446     2.44     0.57   
3    62   317  379  455  142   38  295  178  2023  1872     1.54     0.48   
4   134   326  460  171   52   32  175  205  1920  1553     0.98     0.30   

   season_start_year  
0               2012  
1               2012  
2               2012  
3               2012  
4               2012

As you can see the columns **season_start_year**, was sucessfully created

Note: The **New Orleans Hornets (NOH)** were sold to Tom Benson on **April 13, 2012.**

After purchasing the team, Benson indicated that he wanted **to change the club's nickname** to something more regionally appropriate.

On **December 14, 2012**, it was reported that the Hornets would change their nickname to the **New Orleans Pelicans.**

In [17]:
# Checking the array of the NBA teams by applying the .unique() method,

nba_data.TEAM.unique()

array(['OKC', 'LAL', 'MIA', 'HOU', 'NYK', 'GSW', 'MIL', 'POR', 'TOR',
       'BKN', 'CHA', 'LAC', 'BOS', 'UTA', 'PHI', 'IND', 'SAS', 'ATL',
       'CLE', 'NOH', 'DET', 'CHI', 'SAC', 'DAL', 'DEN', 'MEM', 'PHX',
       'ORL', 'MIN', 'WAS', 'NOP'], dtype=object)

**New Orleans Hornets = (NOH)**

**New Orleans Pelicans = (NOP)**

In [18]:
# Checking the total array of the NBA teams by applying the nunique() method,

nba_data.TEAM.nunique()

31

Create a new standar franchise name for the team New Orleans (NO)

In [19]:
# Standardize the NBA teams.

# Create a new standar franchise name for the team New Orleans

# New Orleans Hornets = (NOH)

# New Orleans Pelicans = (NOP)

nba_data['TEAM'].replace(to_replace=['NOP','NOH'], value='NO', inplace=True)

In [20]:
# Checking the array of the NBA teams by applying the .unique() method,
# with the new New Orleans (NO)

nba_data.TEAM.unique()

array(['OKC', 'LAL', 'MIA', 'HOU', 'NYK', 'GSW', 'MIL', 'POR', 'TOR',
       'BKN', 'CHA', 'LAC', 'BOS', 'UTA', 'PHI', 'IND', 'SAS', 'ATL',
       'CLE', 'NO', 'DET', 'CHI', 'SAC', 'DAL', 'DEN', 'MEM', 'PHX',
       'ORL', 'MIN', 'WAS'], dtype=object)

In [21]:
# Checking the total array of the NBA teams by applying the nunique() method,

nba_data.TEAM.nunique()

30

After standardize the NBA teams.

We can see that now we have **30 teams**

**New Orleans Hornets = (NO)**

In [22]:
# Let's clean and replace the Regular%20Season string that I found on the NBA API url
# Regular Season = RS

nba_data['Season_type'].replace('Regular%20Season','RS', inplace=True)

In [23]:
# Print the first 5 rows of the NBA dataframe by applying the .head() method,
# This method is will display the top 5 observations of the dataset

nba_data.head()

Year Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID TEAM  GP  \
0  2012-13          RS     201142     1     Kevin Durant  1610612760  OKC  81   
1  2012-13          RS        977     2      Kobe Bryant  1610612747  LAL  78   
2  2012-13          RS       2544     3     LeBron James  1610612748  MIA  76   
3  2012-13          RS     201935     4     James Harden  1610612745  HOU  78   
4  2012-13          RS       2546     5  Carmelo Anthony  1610612752  NYK  67   

    MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  DREB  \
0  3119  731  1433   0.510   139   334    0.416  679  750   0.905    46   594   
1  3013  738  1595   0.463   132   407    0.324  525  626   0.839    66   367   
2  2877  765  1354   0.565   103   254    0.406  403  535   0.753    97   513   
3  2985  585  1337   0.438   179   486    0.368  674  792   0.851    62   317   
4  2482  669  1489   0.449   157   414    0.379  425  512   0.830   134   326   

   REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  \
0  640  374  116  105  280  143  2280  2462     1.34     0.41   
1  433  469  106   25  287  173  2133  1921     1.63     0.37   
2  610  551  129   67  226  110  2036  2446     2.44     0.57   
3  379  455  142   38  295  178  2023  1872     1.54     0.48   
4  460  171   52   32  175  205  1920  1553     0.98     0.30   

   season_start_year  
0               2012  
1               2012  
2               2012  
3               2012  
4               2012

Now we can see that the Regular Season = RS

# Creating 2 more data frames:

# 1- Dataframe for Regular Season = regular_season

# 2- Dataframe for Playoffs = playoffs

In [24]:
# Creating Regular Seasons dataframe

regular_season_dataframe = nba_data[nba_data['Season_type']=='RS']


# playoffs_df = data[data['Season_type']=='Playoffs']

In [25]:
# Print the Regular Seasons tables dataframe

regular_season_dataframe

Year Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID TEAM  \
0     2012-13          RS     201142     1     Kevin Durant  1610612760  OKC   
1     2012-13          RS        977     2      Kobe Bryant  1610612747  LAL   
2     2012-13          RS       2544     3     LeBron James  1610612748  MIA   
3     2012-13          RS     201935     4     James Harden  1610612745  HOU   
4     2012-13          RS       2546     5  Carmelo Anthony  1610612752  NYK   
...       ...         ...        ...   ...              ...         ...  ...   
7071  2021-22          RS    1630207   582      Nate Hinton  1610612754  IND   
7072  2021-22          RS    1626155   582       Sam Dekker  1610612761  TOR   
7073  2021-22          RS    1629309   582   Trayvon Palmer  1610612765  DET   
7074  2021-22          RS    1629788   582       Tyler Hall  1610612752  NYK   
7075  2021-22          RS    1629597   582   Zylan Cheatham  1610612762  UTA   

      GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0     81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1     78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2     76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3     78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4     67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   
...   ..   ...  ...   ...     ...   ...   ...      ...  ...  ...     ...   
7071   2     2    0     1   0.000     0     0    0.000    0    0   0.000   
7072   1     1    0     0   0.000     0     0    0.000    0    0   0.000   
7073   1    17    0     1   0.000     0     0    0.000    0    0   0.000   
7074   1     2    0     0   0.000     0     0    0.000    0    0   0.000   
7075   1     5    0     3   0.000     0     2    0.000    0    0   0.000   

      OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  \
0       46   594  640  374  116  105  280  143  2280  2462     1.34     0.41   
1       66   367  433  469  106   25  287  173  2133  1921     1.63     0.37   
2       97   513  610  551  129   67  226  110  2036  2446     2.44     0.57   
3       62   317  379  455  142   38  295  178  2023  1872     1.54     0.48   
4      134   326  460  171   52   32  175  205  1920  1553     0.98     0.30   
...    ...   ...  ...  ...  ...  ...  ...  ...   ...   ...      ...      ...   
7071     0     0    0    0    0    0    1    0     0    -2     0.00     0.00   
7072     0     0    0    0    0    0    0    0     0     0     0.00     0.00   
7073     0     2    2    0    0    0    1    2     0     0     0.00     0.00   
7074     0     0    0    0    0    0    0    0     0     0     0.00     0.00   
7075     0     0    0    0    0    0    0    0     0    -3     0.00     0.00   

      season_start_year  
0                  2012  
1                  2012  
2                  2012  
3                  2012  
4                  2012  
...                 ...  
7071               2021  
7072               2021  
7073               2021  
7074               2021  
7075               2021  

[5148 rows x 31 columns]

As we can see that the Regular Season dataframe contains 5148 rows × 29 columns

In [26]:
# Creating Playoffs dataframe

playoffs_dataframe = nba_data[nba_data['Season_type']=='Playoffs']

In [27]:
# Print the Playoffs dataframe tables dataframe

playoffs_dataframe

Year Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID TEAM  \
468   2012-13    Playoffs       2544     1     LeBron James  1610612748  MIA   
469   2012-13    Playoffs       2225     2      Tony Parker  1610612759  SAS   
470   2012-13    Playoffs       1495     3       Tim Duncan  1610612759  SAS   
471   2012-13    Playoffs     202331     4      Paul George  1610612754  IND   
472   2012-13    Playoffs       2548     5      Dwyane Wade  1610612748  MIA   
...       ...         ...        ...   ...              ...         ...  ...   
7288  2021-22    Playoffs    1629006   206      Josh Okogie  1610612750  MIN   
7289  2021-22    Playoffs    1630556   206  Kessler Edwards  1610612751  BKN   
7290  2021-22    Playoffs    1630201   206    Malachi Flynn  1610612761  TOR   
7291  2021-22    Playoffs     202693   206  Markieff Morris  1610612748  MIA   
7292  2021-22    Playoffs     200794   206     Paul Millsap  1610612755  PHI   

      GP  MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  \
468   23  960  212  432   0.491    36    96    0.375  136  175   0.777    37   
469   21  765  167  365   0.458    11    31    0.355   87  112   0.777    14   
470   21  735  151  321   0.470     0     1    0.000   79   98   0.806    54   
471   19  780  119  277   0.430    34   104    0.327   93  128   0.727    16   
472   22  782  144  315   0.457     1     4    0.250   60   80   0.750    38   
...   ..  ...  ...  ...     ...   ...   ...      ...  ...  ...     ...   ...   
7288   1    2    0    0   0.000     0     0    0.000    0    0   0.000     0   
7289   2    7    0    0   0.000     0     0    0.000    0    0   0.000     0   
7290   6   36    0    7   0.000     0     3    0.000    0    0   0.000     2   
7291   2    3    0    1   0.000     0     0    0.000    0    0   0.000     0   
7292   1    6    0    0   0.000     0     0    0.000    0    0   0.000     0   

      DREB  REB  AST  STL  BLK  TOV  PF  PTS  EFF  AST_TOV  STL_TOV  \
468    156  193  152   41   18   70  43  596  671     2.17     0.59   
469     54   68  146   24    3   53  26  432  397     2.75     0.45   
470    160  214   40   18   34   42  53  381  456     0.95     0.43   
471    125  141   96   25    9   75  72  365  368     1.28     0.33   
472     64  102  105   38   23   58  45  349  368     1.81     0.66   
...    ...  ...  ...  ...  ...  ...  ..  ...  ...      ...      ...   
7288     0    0    0    0    0    0   0    0    0     0.00     0.00   
7289     0    0    1    1    0    1   3    0    1     1.00     1.00   
7290     1    3    3    1    0    1   6    0   -1     3.00     1.00   
7291     1    1    0    0    0    1   2    0   -1     0.00     0.00   
7292     1    1    1    0    0    0   1    0    2     0.00     0.00   

      season_start_year  
468                2012  
469                2012  
470                2012  
471                2012  
472                2012  
...                 ...  
7288               2021  
7289               2021  
7290               2021  
7291               2021  
7292               2021  

[2145 rows x 31 columns]

As we can see that the Regular Season dataframe contains 2145 rows × 29 columns

# NBA Glossary

In order to understand better the NBA dataset columns, here is the Glossary

https://www.nba.com/stats/help/glossary

There are columns that are totals, ratios, or percentages.

Example:

The column called **Field Goals Attempted (FGA)** is a total.

The column called **Field Goal Percentage (FG_PCT)** is the result of two attempts.

I want to create a list of the correct columns to use and conduct a group by.

**Glossary:**

**Year:** The NBA season year.

**Season_type:** Type of the season (e.g., Regular Season).

**PLAYER_ID:** A unique identifier for each player.

**RANK:** Ranking of players based on a specific metric.

**PLAYER:** Player's name.

**TEAM_ID:** A unique identifier for each team.

**TEAM:** Team's abbreviation.

**GP:** Games Played.

**MIN:** Minutes Played.

**FGM, FGA, FG%:** Field Goals Made, Attempted, and Percentage.

**3PM, 3PA, 3P%:** Three-Point Field Goals Made, Attempted, and Percentage.

**FTM, FTA, FT%:** Free Throws Made, Attempted, and Percentage.

**OREB, DREB, REB:** Offensive, Defensive, and Total Rebounds.

**AST:** Assists.

**STL:** Steals.

**BLK:** Blocks.

**TOV:** Turnovers.

**PF:** Personal Fouls.

**PTS:** Points.

**EFF:** Efficiency rating.

**AST_TOV:** Assist to Turnover Ratio.

**STL_TOV:** Steals to Turnover Ratio.

In [28]:
# Creating NBA data col

nba_data.columns

Index(['Year', 'Season_type', 'PLAYER_ID', 'RANK', 'PLAYER', 'TEAM_ID', 'TEAM',
       'GP', 'MIN', 'FGM', 'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT', 'FTM',
       'FTA', 'FT_PCT', 'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK', 'TOV',
       'PF', 'PTS', 'EFF', 'AST_TOV', 'STL_TOV', 'season_start_year'],
      dtype='object')

# Training the model : Feedforward Neural Network (FNN)

## Step 1: Preprocessing

I am going to implement **Feedforward Neural Network (FNN)** to predict player performance efficiency (EFF) based on other statistics. 

The model will predict the efficiency (EFF) of NBA players based on their season statistics. 

## Selecting the features for the model and prepare the data.

In [32]:
# Selecting features and target

features = nba_data[['MIN', 'FGM', 'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT', 'FTM', 'FTA', 'FT_PCT', 'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']]
target = nba_data['EFF']

# Splitting the dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)

# Normalizing the features

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Step 2: Building the Model

I am building the **Feedforward Neural Network (FNN)** using TensorFlow and Keras.

In [35]:
# Determine the number of input features

n_features = X_train_scaled.shape[1]

# Building the model

model = Sequential([
    Dense(64, activation='relu', input_shape=(n_features,)),
    Dense(64, activation='relu'),
    Dense(1)  # Output layer
])

# Compile the model

model.compile(optimizer='adam', loss='mean_squared_error')

## Step 3: Training the Model

I am training the model with our training data.

In [37]:
model.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_split=0.1)

Epoch 1/100
165/165 [==============================] - 0s 1ms/step - loss: 315888.3750 - val_loss: 76672.1484
Epoch 2/100
165/165 [==============================] - 0s 419us/step - loss: 30286.3242 - val_loss: 21038.0039
Epoch 3/100
165/165 [==============================] - 0s 423us/step - loss: 17435.3105 - val_loss: 13826.4922
Epoch 4/100
165/165 [==============================] - 0s 419us/step - loss: 11075.1826 - val_loss: 8771.9326
Epoch 5/100
165/165 [==============================] - 0s 418us/step - loss: 7946.4692 - val_loss: 6699.9316
Epoch 6/100
165/165 [==============================] - 0s 410us/step - loss: 6249.1914 - val_loss: 5427.3311
Epoch 7/100
165/165 [==============================] - 0s 410us/step - loss: 4957.4360 - val_loss: 4282.3491
Epoch 8/100
165/165 [==============================] - 0s 416us/step - loss: 3850.9136 - val_loss: 3215.0049
Epoch 9/100
165/165 [==============================] - 0s 412us/step - loss: 2827.9858 - val_loss: 2292.5234
Epoch 10/100


165/165 [==============================] - 0s 505us/step - loss: 15.8464 - val_loss: 13.1008
Epoch 78/100
165/165 [==============================] - 0s 422us/step - loss: 16.4199 - val_loss: 13.8398
Epoch 79/100
165/165 [==============================] - 0s 413us/step - loss: 12.4889 - val_loss: 13.7984
Epoch 80/100
165/165 [==============================] - 0s 411us/step - loss: 17.7682 - val_loss: 22.1714
Epoch 81/100
165/165 [==============================] - 0s 408us/step - loss: 12.2052 - val_loss: 9.2746
Epoch 82/100
165/165 [==============================] - 0s 409us/step - loss: 11.4982 - val_loss: 10.4989
Epoch 83/100
165/165 [==============================] - 0s 408us/step - loss: 11.6587 - val_loss: 19.8758
Epoch 84/100
165/165 [==============================] - 0s 408us/step - loss: 14.1013 - val_loss: 18.5070
Epoch 85/100
165/165 [==============================] - 0s 411us/step - loss: 11.1637 - val_loss: 8.8471
Epoch 86/100
165/165 [==============================] - 0s 40

## Step 4: Evaluating the Model

I am evaluating the model's performance on the test set.

In [38]:
model.evaluate(X_test_scaled, y_test)

46/46 [==============================] - 0s 377us/step - loss: 13.8733


13.873261451721191